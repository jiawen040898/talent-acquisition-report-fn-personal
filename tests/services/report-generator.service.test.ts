import { AppNotificationService, dateTimeUtil } from '@pulsifi/fn';

import {
    FeatureToggleService,
    FileService,
    ReportGeneratorService,
    S3Service,
} from '../../src/services';
import { testData } from './fixtures/report-generator.service.test-data';

jest.mock('../../src/services/template.service', () => {
    const mBInstance = {
        generate: jest.fn(() => 'buffer'),
    };
    const mB = jest.fn(() => mBInstance);
    return { TemplateService: mB };
});

const mockSQSClient = {
    send: jest.fn(),
};

jest.mock('@aws-sdk/client-sqs', () => {
    const actualFn = jest.requireActual('@aws-sdk/client-sqs');
    return {
        ...actualFn,
        SQSClient: jest.fn(() => mockSQSClient),
    };
});

jest.mock('../../src/services/s3.service', () => {
    const mBInstance = {
        uploadItemInStream: jest.fn(() => 'download link'),
        getFileObjectOutput: jest.fn(() => 'buffer'),
        getFileObjectInBuffer: jest.fn(() => 'buffer'),
        listS3Objects: jest.fn(() => [
            {
                Contents: [1],
                IsTruncated: true,
                NextContinuationToken: 'token',
            },
        ]),
    };
    const mB = jest.fn(() => mBInstance);
    return { S3Service: mB };
});

jest.spyOn(dateTimeUtil, 'getEpochNumber').mockImplementation(
    () => 1704873909104,
);

jest.spyOn(
    FileService.prototype,
    'zipAndUploadCandidateReportAndFiles',
).mockResolvedValue('downloadLink');

const mockedIsFeatureOn = jest.fn();
FeatureToggleService.isFeatureFlagOn = mockedIsFeatureOn;

describe('ReportGeneratorService', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('generateReportPDFAndExtractFileThenUpload', () => {
        it('should generate pdf and extract files if any and upload to idempotence folder based on report summary provided', async () => {
            // Arrange
            const reportGeneratorService = new ReportGeneratorService();
            const spyParallelUpload = jest.spyOn(
                new S3Service(),
                'uploadItemInStream',
            );

            // Act
            const actual =
                await reportGeneratorService.generateReportPDFAndExtractFileThenUpload(
                    testData.mockRenderAssessmentReportData,
                );

            // Assert
            expect(actual).toMatchSnapshot();
            expect(spyParallelUpload.mock.calls).toMatchSnapshot();
        });
    });

    describe('zipReportPDFAndFilesAndSendNotification', () => {
        it('should zip files and send email with the download link', async () => {
            // Arrange
            const reportGeneratorService = new ReportGeneratorService();

            // Act
            await reportGeneratorService.zipReportPDFAndFilesAndSendNotification(
                testData.mockRenderAssessmentReportResultDataToEmail,
            );

            // Assert
            expect(mockSQSClient.send.mock.calls).toMatchSnapshot();
        });

        it('should zip files and send in app message with the download link', async () => {
            // Arrange
            const notificationSend = jest.fn();
            AppNotificationService.notifyUser = notificationSend;

            const reportGeneratorService = new ReportGeneratorService();

            // Act
            await reportGeneratorService.zipReportPDFAndFilesAndSendNotification(
                testData.mockRenderAssessmentReportResultDataToInApp,
            );

            // Assert
            expect(notificationSend.mock.calls).toMatchSnapshot();
        });
    });
});
