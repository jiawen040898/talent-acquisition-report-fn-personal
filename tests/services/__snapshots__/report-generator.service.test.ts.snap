// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportGeneratorService generateReportPDFAndExtractFileThenUpload should generate pdf and extract files if any and upload to idempotence folder based on report summary provided 1`] = `
{
  "count": 1,
  "folderPath": "report/assessment/5/00000000-0000-0000-0000-000000000001",
  "variable": {
    "company_id": 5,
    "locale": "en-US",
    "notification_type": "email",
    "timezone": "Asia/Kuala_Lumpur",
    "user_account_id": 1,
    "user_email": "jaypete@gmail.com",
    "user_first_name": "Jay",
  },
}
`;

exports[`ReportGeneratorService generateReportPDFAndExtractFileThenUpload should generate pdf and extract files if any and upload to idempotence folder based on report summary provided 2`] = `
[
  [
    {
      "Body": "buffer",
      "Bucket": "pulsifi-sandbox-document-download",
      "Key": "report/assessment/5/00000000-0000-0000-0000-000000000001/2. Simon Peter/Simon Peter - Report.pdf",
    },
  ],
  [
    {
      "Body": undefined,
      "Bucket": "pulsifi-sandbox-document-download",
      "Key": "report/assessment/5/00000000-0000-0000-0000-000000000001/2. Simon Peter/Simon Peter Resume.pdf",
    },
  ],
  [
    {
      "Body": undefined,
      "Bucket": "pulsifi-sandbox-document-download",
      "Key": "report/assessment/5/00000000-0000-0000-0000-000000000001/2. Simon Peter/Simon Peter Attachment-0.pdf",
    },
  ],
]
`;

exports[`ReportGeneratorService zipReportPDFAndFilesAndSendNotification should zip files and send email with the download link 1`] = `
[
  [
    SendMessageCommand {
      "input": {
        "MessageBody": "{"recipient_email":"jaypete@gmail.com","recipient_id":"1","recipient_group":"user","email_communication_type":"application_document_download_confirmation","company_id":5,"variables":{"first_name":"Jay","download_link":"downloadLink"}}",
        "QueueUrl": "https://test.pulsifi.me/queue/notitication/email",
      },
      "middlewareStack": {
        "add": [Function],
        "addRelativeTo": [Function],
        "applyToStack": [Function],
        "clone": [Function],
        "concat": [Function],
        "identify": [Function],
        "identifyOnResolve": [Function],
        "remove": [Function],
        "removeByTag": [Function],
        "resolve": [Function],
        "use": [Function],
      },
    },
  ],
]
`;

exports[`ReportGeneratorService zipReportPDFAndFilesAndSendNotification should zip files and send in app message with the download link 1`] = `
[
  [
    5,
    "1",
    {
      "event_data": {
        "data": {
          "download_link": "downloadLink",
        },
        "meta": {
          "company_id": 5,
          "correlation_id": "00000000-0000-0000-0000-000000000003",
          "env": "test",
          "user_account_id": 1,
        },
      },
      "event_type": "report-download-ready",
    },
  ],
]
`;
